### ISSUE TITLE
[1.8.0.beta2] vim mode shows only empty editor pane

### ISSUE THREAD

**Issue by TS-CUBED** [Zettlr/Zettlr#1415]
## Description
vim mode in new beta2 is broken (also in 1.8.0 alpha, beta, and beta1)

## Reproducing
- start Zettlr 1.8.0beta2 with empty config dir
- open a directory with zettlr notes
- select a note -> editor shows note
- go to settings - Editor, select "Editor Input Mode" = "Vim"
- close and restart Zettlr
- select a note -> note opens (tab is displayed), but editor shows blank pane

## Expected behaviour
- note should open in editor and vim mode active

## Platform
 - OS and version: Ubuntu 20.04 LTS
 - Zettlr Version: 1.8.0 beta2

**Comment by TS-CUBED** [comment]
bug also manifests without restart:
- open note with editor input mode "Normal"
- note displays
- click in window allows note to be edited

- switch to editor input mode "Vim"
- click in window does not show cursor, note cannot be edited

**Comment by nathanlesage** [comment]
I've noted myself that the vim-mode exhibits some bugs. As stated on the [website](https://codemirror.net/demo/vim.html):

> Note: The CodeMirror vim bindings do not have an active maintainer. That means that if you report bugs in it, they are likely to go unanswered.

This means I am pretty sure I cannot help you. I'm just loading the vim keybindings and using them.

**Comment by TS-CUBED** [comment]
The problem is only present in the 1.8 branch. It works fine in 1.7.5.
If the problem was with CodeMirror, wouldn't it show up in both versions?

**Comment by nathanlesage** [comment]
There have been updates to CodeMirror as well, so this might also be a reason.

**Comment by TS-CUBED** [comment]
Hmm, interesting. Based on your last comment I played around with this a bit. I am no JS developer, so take all of this with a grain of salt:

- cloned the Zettlr repo
- checked out 1.8.0-beta.2
- rolled back codemirror to version 5.56.0 (2020-07-20) - which is the same as used in v1.7.5
- yarn install, yarn start
- same problem - works without vim keybindings, but not with them.

**Comment by TS-CUBED** [comment]
- and on the other hand v1.7.5 works, if I compile it with codemirror 5.58.2 - which is the one v1.8.0-beta.2 will compile by default.

**Comment by TS-CUBED** [comment]
There seem to be a few problems with the package dependencies in 1.8.0 - could that be related?
```
[3/4] Linking dependencies...
warning " > @clr/icons@4.0.4" has unmet peer dependency "@webcomponents/custom-elements@^1.0.0".
warning " > copy-webpack-plugin@6.1.0" has unmet peer dependency "webpack@^4.37.0 || ^5.0.0".
warning " > css-loader@3.6.0" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > eslint-loader@4.0.2" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > file-loader@6.1.1" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > less-loader@7.0.2" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > node-loader@1.0.2" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > raw-loader@4.0.2" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > style-loader@1.2.1" has unmet peer dependency "webpack@^4.0.0 || ^5.0.0".
warning " > vue-loader@15.9.3" has unmet peer dependency "webpack@^3.0.0 || ^4.1.0 || ^5.0.0-0".
```

**Comment by KarlClinckspoor** [comment]
Can confirm that this has happened to me as well. Needs to change the editing mode to Normal and restart the application. Emacs mode doesn't have this problem. Maybe the Vim mode entry should be removed until this issue is fixed in CoreMirror, for new users' sake? If someone wants to test, they change the mode like it was done before the dropdown list was added.

**Comment by nathanlesage** [comment]
> Maybe the Vim mode entry should be removed until this issue is fixed in CoreMirror, for new users' sake?

I agree. The Vim mode right now doesn't have a maintainer, and I'm very much in favor of discontinuing the feature until someone has been found.

**Comment by bit-hug** [comment]
NB: This bug is now live in the stable release, and possibly quite scary for some people updating to suddenly see empty files in place of their previously open documents.

**Comment by bit-hug** [comment]
Also noted that @ianhi offered assistance with codemirror vim issues.

**Comment by nathanlesage** [comment]
> This bug is now live in the stable release, and possibly quite scary for some people updating from 1.7.5 to suddenly see empty files in place of their previously open documents.

You are absolutely right, and I should quickly discourage use of the VIM mode …

**Comment by bit-hug** [comment]
Maybe a comment in the release notes would be useful. I'm already missing vim mode, it's a really good feature even though it doesn't work perfectly...

**Comment by lukeallister** [comment]
For me, loss of vim mode is a dealbreaker. I'll be sticking with 1.7.5 until this is resolved.

**Comment by TS-CUBED** [comment]
Same here. I've moved on to org-mode, though.

The irony, moving back to emacs after over 20 years because my
Zettelkasten lost VIM keybindings ;-)

**Comment by lukeallister** [comment]
@TS-CUBED Too funny. 

Do you have any sense of what's going on with Vim-mode, or where to look to try to fix it? I'm not a js developer (and have to finish grad school before starting any extra projects) but I might try to dig into this if we can't find anyone who actually knows what they're doing.

**Comment by TS-CUBED** [comment]
Not really. I'm also no JS developer.

I cloned and compiled 1.8 beta 2 with a rolled back version of
codemirror (the same as in 1.7.5) and the bug is there as well. So I
guess the bug is in Zettlr rather than codemirror.

The VIM bindings seem to work in other codemirror/electron apps as well.

But I don't really use Zettlr anymore - but my wife does, albeit without
VIM mode. I had already moved to NVIM with vim-wiki and vim-zettel
(which is compatible with Zettlr) and now to emacs with org-mode. But
that is a completely different beast.

**Comment by nathanlesage** [comment]
> So I guess the bug is in Zettlr rather than codemirror.

That's an important thing. I'll have a look some time. But please know that VIM mode has no real priority, as the main target audience of Zettlr is non-technical folk either way, and I have worked with vim myself, and it's really only useful on the command line, not in a GUI application. That is subjective, sure, but the way it is here …

**Comment by lukeallister** [comment]
> [Vim] is really only useful on the command line

That is a disappointing assessment. Obviously you need to focus on the features that you think are most important, but I would like to point out the extra value in combining markdown with Vim (or emacs). 

The point of markdown is to be able to _type_ your formatting instead of switching to the mouse. The point of Vim is to make text navigation/editing as efficient as possible. Those two things stack really nicely together, saving a lot of time. On the other side, I do not think that non-technical folks have much interest in markdown anyway. Non-technical users people will gravitate towards WYSIWYG editors, happy to use the mouse to select formatting. 

Ultimately you should do what you think is best, but I think that there is a significant intersection between markdown users and technical users.

**Comment by nathanlesage** [comment]
> On the other side, I do not think that non-technical folks have much interest in markdown anyway. Non-technical users people will gravitate towards WYSIWYG editors, happy to use the mouse to select formatting.

This is what I intend to _change_ – and to do so is to increase the intersection between Markdown users and non-technical users, do you understand what my intention with the selection of features for me is? My aim is to get non-technical folk to get to Markdown, because certainly, the intersection between Markdown and tech is large, but my intention is to increase the intersection between Markdown and non-tech. I don't take things just as they are, but I want to change the things as they are!

**Comment by sj4nes** [comment]
This was a surprising regression to me, as having previously set Vi mode.  I had to come here to look and see if I missed something about the upgrade.

**Comment by nathanlesage** [comment]
@sj4nes This indicates that the vim bug apparently imparts saving of the settings, which should not be. The negative numbers are hashes. I've re-implemented the Java hashing function, and this sometimes gives negative, sometimes positive numbers. Everything to be expected!

**Comment by nathanlesage** [comment]
There's just a new issue open, whose title is better for others who might stumble upon this, so I'll close this one for the other one, please head over here: #1473

**Comment by TS-CUBED** [comment]
> > So I guess the bug is in Zettlr rather than codemirror.
> 
> That's an important thing. I'll have a look some time. But please know that VIM mode has no real priority, as the main target audience of Zettlr is non-technical folk either way, and I have worked with vim myself, and it's really only useful on the command line, not in a GUI application. That is subjective, sure, but the way it is here …

I wrote that in another message on this thread.

BTW, I must object to the statement that VIM mode is only for technical users. The modal editor concept, once mastered, is a skill that will benefit most touch-typists, in particular in the humanities. It's a bit like markdown languages. They may have originated in a more technical context, but can be beneficial to anybody who needs to write. Your statement is just perpetuating the paradigm that you need a mouse if you are editing text in a GUI program, which I'd say, is incorrect.

BTW2: I have introduced my wife (PhD student in Archaeology/Heritage management) to markdown (with Zettlr). And the one thing that would help adoption there is a preview mode, since her non-technical background causes a preference for visual confirmation/prediction of the layout. A feature that you have robustly argued against.

I agree with you that that feature should not be required, but I'd say that that is biased by our technical background, and it is only through my influence that she has not dropped Zettlr yet due to the lack of a preview pane.

**Comment by bit-hug** [comment]
Just to note, the editor is not in fact frozen under this bug, and will silently accept input and save to file, but this will not be displayed until Zettlr is restarted in normal mode. Perhaps the title of this issue is therefore more accurate than the title of issue #1473

**Comment by nathanlesage** [comment]
@bit-hug I see your objection, and for me it's more about a problem that it feels as if VIM mode editing belongs more to the command line than a GUI, but then I'm just doing this in my free time and don't have that much mental capacity to actually conduct studies and weigh things in order to deliver a good experience for everyone. My rejection of VIM mode is more based upon the fact that I need to somehow cut my work because there's so much that's still buggy, slow or not optimized, and I want to prevent myself getting burnt out. This is why I'm cutting all the purely technological stuff, it's simply a heuristic I'm applying.

As mentioned in #1473 I'll gladly accept PRs introducing such functionality, but only (!) if I do not have to maintain it. You know what I'm getting at? I don't want to disappoint anyone, but given the fact that I am right now working on the app for my whole Sunday, because I don't have time under the week, this is actually a warning sign.

**Comment by TS-CUBED** [comment]
Totally understand that. Zettlr is a great app, and definitely a huge
step towards getting markdown and Zettelkasten out to the less technical
people who can benefit from it.

Thanks for all the hard work you put in it.

**Comment by bit-hug** [comment]
> @bit-hug I see your objection, and for me it's more about a problem that it feels as if VIM mode editing belongs more to the command line than a GUI, but then I'm just doing this in my free time and don't have that much mental capacity to actually conduct studies and weigh things in order to deliver a good experience for everyone. My rejection of VIM mode is more based upon the fact that I need to somehow cut my work because there's so much that's still buggy, slow or not optimized, and I want to prevent myself getting burnt out. This is why I'm cutting all the purely technological stuff, it's simply a heuristic I'm applying.
> 
> As mentioned in #1473 I'll gladly accept PRs introducing such functionality, but only (!) if I do not have to maintain it. You know what I'm getting at? I don't want to disappoint anyone, but given the fact that I am right now working on the app for my whole Sunday, because I don't have time under the week, this is actually a warning sign.

That's very understandable.

I actually wish I had more that I could offer, but any technical knowledge or skill I have in programming is minimal and only comes from using various tools in my physics work. I hope there comes a time when I can make a meaningful contribution.

As for the utility of vim mode, I don't use vim independently very much, but I simply find the keybindings and associated small functions very convenient and efficient. It's a big bonus for me if Zettlr manages to incorporate it again in the future, but it was only one of very many good things about Zettlr.

Broadly speaking, in my view, the goals of enabling people from a less technical background to use markdown, and enabling people from a less technical background to use command line tools, are harmonious and mutually aligned ones. That you have done and continue to do so much to further even one of these aims (and not in a small way) is already to be very highly commended. More power to you!

**Comment by nathanlesage** [comment]
@bit-hug Thank you very much for your kind words, and I appreciate your understanding. I really hope that with rising popularity of the app some helping hands will find their way in here and we can resume implementing those features as well :)
